crates/bevy_ecs/src/bundle/info.rs

```diff
@@ -111,41 +112,106 @@ impl BundleInfo {
+        // - is in the same order as the source bundle type
+        let mut info = BundleInfo {
+            id,
+            component_ids,
+            required_components: Vec::new(),
+            explicit_components_len,
+        };
+
+        // SAFETY: first call with the same `storages` and `components`.
+        unsafe {
+            info.set_required_components(storages, components, components_in_bundles, |_| true);
+        }
+
+        info
+    }
+
+    /// Update [`Self::component_ids`] to contain all required components, updates [`Self::required_components`] alongside.
+    ///
+    /// The filter `components_in_bundles_filter` determines for which components [`Self::id`] is added to `components_in_bundles`.
+    ///
+    /// # Safety
+    ///
+    /// `Self` must be constructed from [`Self::new`] with the same `storages` and `components`.
+    unsafe fn set_required_components(
+        &mut self,
+        storages: &mut Storages,
+        components: &Components,
+        components_in_bundles: &mut Vec<Vec<BundleId>>,
+        components_in_bundles_filter: impl Fn(ComponentId) -> bool,
+    ) {
+        let mut components_in_bundles_push = |component: ComponentId| {
+            if !components_in_bundles_filter(component) {
+                return;
+            }
+            match components_in_bundles.get_mut(component.index()) {
+                Some(bundles) => bundles.push(self.id),
+                None => {
+                    let missing_minus_one = component.index() - components_in_bundles.len();
+                    let iter = core::iter::repeat_n(Vec::new(), missing_minus_one)
+                        .chain(core::iter::once(vec![self.id]));
+                    components_in_bundles.reserve_exact(missing_minus_one + 1);
+                    components_in_bundles.extend(iter);
+                }
+            }
+        };
+
+        // handle explicit components
         let mut required_components = RequiredComponents::default();
-        for component_id in component_ids.iter().copied() {
-            // SAFETY: caller has verified that all ids are valid
+        for component_id in self.component_ids.iter().copied() {
+            // SAFETY: caller of initial constructor has verified that all ids are valid
             let info = unsafe { components.get_info_unchecked(component_id) };
             required_components.merge(info.required_components());
             storages.prepare_component(info);
+            components_in_bundles_push(component_id);
         }
-        required_components.remove_explicit_components(&component_ids);
+        required_components.remove_explicit_components(&self.component_ids);
 
         // handle required components
-        let required_components = required_components
-            .0
-            .into_iter()
-            .map(|(component_id, v)| {
+        self.required_components
+            .extend(required_components.0.into_iter().map(|(component_id, v)| {
```

Comment:
This `extend` seems fishy. I can see that this method is always called when `self.required_components` is empty, and this might make sense for reusing its storage, but the invariant is not documented anywhere. I would instead call `self.required_components.clear()` here and document that we are doing so to reuse its allocation (not that this should really matter, I wouldn't expect people to register required components so often that this starts becoming a bottleneck).